### PR TITLE
Kamila/set score game

### PR DESCRIPTION
**Pull request by kamilababayeva**
This PR updates the score after each typing game ended based on genre/mode.

Changes:
- mock typing game
- add functions to set score per each genre in DataGetter
- save mode/genre in GameManager

TODO next pr: save score per lyrics game, update UI

### CODE REVIEW

File: app/src/main/java/ch/sdp/vibester/database/DataGetter.kt [hunk 1]
```diff
@@ -49,45 +49,58 @@ class DataGetter @Inject constructor() {
...
-                .get().addOnSuccessListener { t ->
-                    if(!TestMode.isTest()) {
-                        updateFieldInt(userID, (t.value as Long?)!!.toInt() + newVal, fieldName)
-                    }
-                }
-        }
+        dbUserRef.child(userID).child(fieldName)
+            .get().addOnSuccessListener { t ->
+                var sum  = newVal
+                if(t.value != null){sum += (t.value as Long?)!!.toInt()}
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/database/DataGetter.kt:60]:
```suggestion
                if(t.value != null) {sum += (t.value as Long?)!!.toInt()}
```

File: app/src/main/java/ch/sdp/vibester/database/DataGetter.kt [hunk 2]
```diff
@@ -49,45 +49,58 @@ class DataGetter @Inject constructor() {
...
-        if(!TestMode.isTest()) {
-            dbUserRef.child(userID).child(fieldName)
-                .get().addOnSuccessListener { t ->
-                    if(!TestMode.isTest()) {
-                        updateFieldInt(userID, maxOf((t.value as Long?)!!.toInt(), newVal), fieldName)
-                    }
+        dbUserRef.child(userID).child(fieldName)
+            .get().addOnSuccessListener { t ->
+                var maxValue = newVal
+                if(t.value != null){maxValue = maxOf((t.value as Long?)!!.toInt(), newVal)}
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/database/DataGetter.kt:76]:
```suggestion
                if(t.value != null) {maxValue = maxOf((t.value as Long?)!!.toInt(), newVal)}
```

File: app/src/main/java/ch/sdp/vibester/database/DataGetter.kt [hunk 3]
```diff
@@ -49,45 +49,58 @@ class DataGetter @Inject constructor() {
...
+     * @param userID
+     * @param newVal integer value to update
+     * @param fieldName field name to updated
+     * @param subFieldName subfield name to updated
+     */
+    fun updateBestSubFieldInt(userID: String, newVal: Int, fieldName: String, subFieldName: String) {
+        dbUserRef.child(userID).child(fieldName).child(subFieldName)
+            .get().addOnSuccessListener { t ->
+                var maxValue = newVal
+                if(t.value != null){maxValue = maxOf((t.value as Long?)!!.toInt(), newVal)}
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/database/DataGetter.kt:92]:
```suggestion
                if(t.value != null) {maxValue = maxOf((t.value as Long?)!!.toInt(), newVal)}
```

File: app/src/main/java/ch/sdp/vibester/database/DataGetter.kt [hunk 4]
```diff
@@ -49,45 +49,58 @@ class DataGetter @Inject constructor() {
...
+    }
+
+    /**
+     * Update the best value of subfield
+     * @param userID
+     * @param newVal integer value to update
+     * @param fieldName field name to updated
+     * @param subFieldName subfield name to updated
+     */
+    fun updateBestSubFieldInt(userID: String, newVal: Int, fieldName: String, subFieldName: String) {
```
Comment (jiabaow) [inline app/src/main/java/ch/sdp/vibester/database/DataGetter.kt:88]: The three update functions are very similar, refactoring them can not only fix code climate issues but also boost the test coverage. Do you think is doable?

**Review by jiabaow — COMMENTED** [review]
Thanks for this update!
As explained, test coverage is low because of the database which is not a problem, however, do you think it's worth it to refactor the duplicates?

File: app/src/main/java/ch/sdp/vibester/database/DataGetter.kt (hunk 4 above)
Comment (kamilababayeva) [reply app/src/main/java/ch/sdp/vibester/database/DataGetter.kt]: I have refactored the database calls. They have become cleaner, the coverage remained same, and did not fall!

File: app/src/androidTest/java/ch/sdp/vibester/activity/TypingGameActivityTest.kt [hunk 5]
```diff
@@ -46,6 +55,17 @@ class TypingGameActivityTest {
...
 
+    @BindValue
+    @JvmField
+    val mockUsersRepo = mockk<DataGetter>()
+
+    private fun createMockInvocation() {
+        every {mockUsersRepo.setSubFieldValue(any(), any(), any(),any())} answers {}
+        every {mockUsersRepo.updateFieldInt(any(), any(), any(), any())} answers {}
+        every {mockUsersRepo.setFieldValue(any(), any(), any())} answers {}
+        every {mockUsersRepo.updateSubFieldInt(any(), any(), any(), any(), any())} answers {}
```
Comment (MaximeZmt) [inline app/src/androidTest/java/ch/sdp/vibester/activity/TypingGameActivityTest.kt:66]:
```suggestion
        every { mockUsersRepo.setSubFieldValue(any(), any(), any(), any()) } answers {}
        every { mockUsersRepo.updateFieldInt(any(), any(), any(), any()) } answers {}
        every { mockUsersRepo.setFieldValue(any(), any(), any()) } answers {}
        every { mockUsersRepo.updateSubFieldInt(any(), any(), any(), any(), any()) } answers {}

```

File: app/src/main/java/ch/sdp/vibester/activity/ProfileActivity.kt [hunk 6]
```diff
@@ -101,7 +101,7 @@ class ProfileActivity : AppCompatActivity() {
             findViewById<TextView>(textId).text = input.text.toString()
 
             if(name == "username"){
-                dataGetter.updateFieldString(FireBaseAuthenticator.getCurrentUID(), input.text.toString(), "username")
+                dataGetter.setFieldValue(FireBaseAuthenticator.getCurrentUID(),"username",  input.text.toString())
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/activity/ProfileActivity.kt:104]:
```suggestion
                dataGetter.setFieldValue(FireBaseAuthenticator.getCurrentUID(), "username",  input.text.toString())

```

File: app/src/main/java/ch/sdp/vibester/database/DataGetter.kt [hunk 7]
```diff
@@ -21,81 +21,78 @@ class DataGetter @Inject constructor() {
...
-    fun updateBestFieldInt(userID: String, newVal: Int, fieldName: String) {
-        if(!TestMode.isTest()) {
-            dbUserRef.child(userID).child(fieldName)
-                .get().addOnSuccessListener { t ->
-                    if(!TestMode.isTest()) {
-                        updateFieldInt(userID, maxOf((t.value as Long?)!!.toInt(), newVal), fieldName)
+    fun updateSubFieldInt(userID: String, newVal: Int, fieldName: String, subFieldName: String, method: String) {
+        dbUserRef.child(userID).child(fieldName).child(subFieldName)
+            .get().addOnSuccessListener { t ->
+                var finalVal  = newVal
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/database/DataGetter.kt:83]:
```suggestion
                var finalVal = newVal
```

**Review by MaximeZmt — COMMENTED** [review]
nice work 👍

**Review by MaximeZmt — APPROVED** [review]
👍

**Review by jiabaow — APPROVED** [review]
lgtm

**Pull request merged**